src/lib.rs

```diff
@@ -481,67 +761,163 @@ impl<I: Interchange> Responder<I> {
+                    }
+                    f(i.rp_mut())
+                })
+            };
+            Ok(res)
         } else {
-            None
+            Err(Error)
         }
     }
 
-    /// Send a response that was already placed in the interchange using `response_mut`.
-    pub fn send_response(&mut self) -> Result<(), ()> {
-        if State::BuildingResponse == self.state.load(Ordering::Acquire) {
+    /// Initialize a response with its default values and and return a mutable reference to it
+    ///
+    /// This is usefull to build large structures in-place
+    // Safety: We cannot test this with loom efficiently, but given that `with_response_mut` is tested,
+    // this is likely correct
+    #[cfg(not(loom))]
+    pub fn response_mut(&mut self) -> Result<&mut A, Error> {
+        if self.transition(State::Requested, State::BuildingResponse)
+            || self.transition(State::BuildingResponse, State::BuildingResponse)
+        {
             unsafe {
-                *self.interchange_mut() = I::from_rp(self.response_mut().unwrap());
-            }
-            if self.transition(State::BuildingResponse, State::Responded) {
-                Ok(())
-            } else {
-                Err(())
+                self.with_data_mut(|i| {
+                    if !i.is_response_state() {
+                        *i = Message::from_rp(A::default());
+                    }
+                })
             }
+            Ok(unsafe { self.data_mut().rp_mut() })
+        } else {
+            Err(Error)
+        }
+    }
+
+    /// Send a response that was already placed in the channel using `response_mut` or
+    /// `with_response_mut`.
+    pub fn send_response(&mut self) -> Result<(), Error> {
+        if State::BuildingResponse == self.channel.state.load(Ordering::Acquire)
+            && self.transition(State::BuildingResponse, State::Responded)
+        {
+            Ok(())
         } else {
             // logic error
-            Err(())
+            Err(Error)
+        }
+    }
+}
+
+unsafe impl<'i, Q, A> Send for Responder<'i, Q, A> {}
+unsafe impl<'i, Q, A> Send for Requester<'i, Q, A> {}
+unsafe impl<Q, A> Send for Channel<Q, A> {}
+unsafe impl<Q, A> Sync for Channel<Q, A> {}
+
+/// Set of `N` channels
+///
+/// Channels can be claimed with [`claim()`](Self::claim)
+///
+/// ```
+/// # #![cfg(not(loom))]
+/// # use interchange::*;
+/// # #[derive(Clone, Debug, PartialEq)]
+/// # pub enum Request {
+/// #     This(u8, u32),
+/// #     That(i64),
+/// # }
+/// #
+/// # #[derive(Clone, Debug, PartialEq)]
+/// # pub enum Response {
+/// #     Here(u8, u8, u8),
+/// #     There(i16),
+/// # }
+/// #
+/// static interchange: Interchange<Request, Response,10> = Interchange::new();
+///
+/// for i in 0..10 {
+///     let rq: Requester<'_, Request, Response>;
+///     let rp: Responder<'_, Request, Response>;
+///     (rq, rp) = interchange.claim().unwrap() ;
+/// }
+/// ```
+pub struct Interchange<Q, A, const N: usize> {
```

Comment:
I presume Q = query/question, and A = answer? Maybe use longer names in the struct definition (and switch to the short ones in the implementation)? In any case, please add a comment so it's easiert to understand.

Reply:
You're right, I'm not very satisfied with Q,A since everything else mentions Request/Response. I'll replace it with `Rq` and `Rp`